File: go/test/entity_set_push_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/roadie-sdk/go"
	"github.com/voxgig-sdk/roadie-sdk/go/core"

	vs "github.com/voxgig-sdk/roadie-sdk/go/utility/struct"
)

func TestEntitySetPushEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.EntitySetPush(nil)
		if ent == nil {
			t.Fatal("expected non-nil EntitySetPushEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := entity_set_pushBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "entity_set_push." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set ROADIE_TEST_ENTITY_SET_PUSH_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		entitySetPushRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.entity_set_push", setup.data)))
		var entitySetPushRef01Data map[string]any
		if len(entitySetPushRef01DataRaw) > 0 {
			entitySetPushRef01Data = core.ToMapAny(entitySetPushRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = entitySetPushRef01Data

		// UPDATE
		entitySetPushRef01Ent := client.EntitySetPush(nil)
		entitySetPushRef01DataUp0Up := map[string]any{
		}

		entitySetPushRef01MarkdefUp0Name := "set"
		entitySetPushRef01MarkdefUp0Value := fmt.Sprintf("Mark01-entity_set_push_ref01_%d", setup.now)
		entitySetPushRef01DataUp0Up[entitySetPushRef01MarkdefUp0Name] = entitySetPushRef01MarkdefUp0Value

		entitySetPushRef01ResdataUp0Result, err := entitySetPushRef01Ent.Update(entitySetPushRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		entitySetPushRef01ResdataUp0 := core.ToMapAny(entitySetPushRef01ResdataUp0Result)
		if entitySetPushRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if entitySetPushRef01ResdataUp0[entitySetPushRef01MarkdefUp0Name] != entitySetPushRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", entitySetPushRef01MarkdefUp0Name, entitySetPushRef01ResdataUp0[entitySetPushRef01MarkdefUp0Name])
		}

	})
}

func entity_set_pushBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "entity_set_push", "EntitySetPushTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read entity_set_push test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse entity_set_push test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"entity_set_push01", "entity_set_push02", "entity_set_push03", "set01", "set02", "set03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("ROADIE_TEST_ENTITY_SET_PUSH_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"ROADIE_TEST_ENTITY_SET_PUSH_ENTID": idmap,
		"ROADIE_TEST_LIVE":      "FALSE",
		"ROADIE_TEST_EXPLAIN":   "FALSE",
		"ROADIE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["ROADIE_TEST_ENTITY_SET_PUSH_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["ROADIE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["ROADIE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewRoadieSDK(core.ToMapAny(mergedOpts))
	}

	live := env["ROADIE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["ROADIE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/entity_set_push_entity_test.lua
-- EntitySetPush entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("roadie_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("EntitySetPushEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:EntitySetPush(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = entity_set_push_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"update"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "entity_set_push." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set ROADIE_TEST_ENTITY_SET_PUSH_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local entity_set_push_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.entity_set_push")))
    local entity_set_push_ref01_data = nil
    if #entity_set_push_ref01_data_raw > 0 then
      entity_set_push_ref01_data = helpers.to_map(entity_set_push_ref01_data_raw[1][2])
    end

    -- UPDATE
    local entity_set_push_ref01_ent = client:EntitySetPush(nil)
    local entity_set_push_ref01_data_up0_up = {
    }

    local entity_set_push_ref01_markdef_up0_name = "set"
    local entity_set_push_ref01_markdef_up0_value = "Mark01-entity_set_push_ref01_" .. tostring(setup.now)
    entity_set_push_ref01_data_up0_up[entity_set_push_ref01_markdef_up0_name] = entity_set_push_ref01_markdef_up0_value

    local entity_set_push_ref01_resdata_up0_result, err = entity_set_push_ref01_ent:update(entity_set_push_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local entity_set_push_ref01_resdata_up0 = helpers.to_map(entity_set_push_ref01_resdata_up0_result)
    assert.is_not_nil(entity_set_push_ref01_resdata_up0)
    assert.are.equal(entity_set_push_ref01_resdata_up0[entity_set_push_ref01_markdef_up0_name], entity_set_push_ref01_markdef_up0_value)

  end)
end)

function entity_set_push_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/entity_set_push/EntitySetPushTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read entity_set_push test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "entity_set_push01", "entity_set_push02", "entity_set_push03", "set01", "set02", "set03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("ROADIE_TEST_ENTITY_SET_PUSH_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["ROADIE_TEST_ENTITY_SET_PUSH_ENTID"] = idmap,
    ["ROADIE_TEST_LIVE"] = "FALSE",
    ["ROADIE_TEST_EXPLAIN"] = "FALSE",
    ["ROADIE_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["ROADIE_TEST_ENTITY_SET_PUSH_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["ROADIE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["ROADIE_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["ROADIE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["ROADIE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
